src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,478 @@
+    /// 
+    /// Issue #8193 tracks filling this out to the full specification. 
+    /// 
+    ///     https://github.com/dotnet/roslyn/issues/8193
+    /// </summary>
+    /// <remarks>
+    /// Options which can cause compilation failure, but doesn't impact the result of a successful
+    /// compilation should be included. That is because it is interesting to describe error states
+    /// not just success states. Think about caching build failures as well as build successes.
+    ///
+    /// When an option is omitted, say if there is no value for a public crypto key, we should emit
+    /// the property with a null value vs. omitting the property. Either approach would produce 
+    /// correct results the preference is to be declarative that an option is omitted.
+    /// </remarks>
+    internal abstract class DeterministicKeyBuilder
+    {
+        protected DeterministicKeyBuilder()
+        {
+        }
+
+        protected void WriteFileName(JsonWriter writer, string name, string? filePath, DeterministicKeyOptions options)
+        {
+            if ((options & DeterministicKeyOptions.IgnorePaths) != 0)
+            {
+                filePath = Path.GetFileName(filePath);
+            }
+
+            writer.Write(name, filePath);
+        }
+
+        protected void WriteByteArrayValue(JsonWriter writer, string name, ImmutableArray<byte> value)
+        {
+            if (!value.IsDefault)
+            {
+                WriteByteArrayValue(writer, name, value.AsSpan());
+            }
+        }
+
+        internal static void EncodeByteArrayValue(ReadOnlySpan<byte> value, StringBuilder builder)
+        {
+            foreach (var b in value)
+            {
+                builder.Append(b.ToString("x"));
+            }
+        }
+
+        protected void WriteByteArrayValue(JsonWriter writer, string name, ReadOnlySpan<byte> value)
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            EncodeByteArrayValue(value, builder.Builder);
+            writer.Write(name, builder.ToStringAndFree());
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
+        {
+            writer.WriteKey(key);
+            WriteType(writer, type);
+        }
+
+        protected void WriteType(JsonWriter writer, Type? type)
+        {
+            if (type is null)
+            {
+                writer.WriteNull();
+                return;
+            }
+
+            writer.WriteObjectStart();
+            writer.Write("fullName", type.FullName);
+            // Note that the file path to the assembly is deliberately not included here. The file path
+            // of the assembly does not contribute to the output of the program.
+            writer.Write("assemblyName", type.Assembly.FullName);
+            writer.Write("mvid", type.Assembly.ManifestModule.ModuleVersionId.ToString());
```

Comment:
consider an explicit format flag here.  though i'm not sure if that matters for Guid.